lib/active_merchant/billing/gateways/pin.rb

```diff
@@ -149,23 +149,56 @@ def add_capture(post, options)
         post[:capture] = capture != false
       end
 
-      def add_creditcard(post, creditcard)
-        if creditcard.respond_to?(:number)
-          post[:card] ||= {}
+      def add_payment_method(post, payment_method)
+        return unless payment_method
 
+        case payment_method
+        when NetworkTokenizationCreditCard
+          token = get_single_use_token(payment_method, options)
+          post[:payment_source_token] = token
+        when CreditCard
+          post[:card] ||= {}
           post[:card].merge!(
-            number: creditcard.number,
-            expiry_month: creditcard.month,
-            expiry_year: creditcard.year,
-            cvc: creditcard.verification_value,
-            name: creditcard.name
+            number: payment_method.number,
+            expiry_month: payment_method.month,
+            expiry_year: payment_method.year,
+            cvc: payment_method.verification_value,
+            name: payment_method.name
           )
-        elsif creditcard.kind_of?(String)
-          if /^card_/.match?(creditcard)
-            post[:card_token] = get_card_token(creditcard)
+        when String
+          if /^card_/.match?(payment_method)
+            post[:card_token] = get_card_token(payment_method)
           else
-            post[:customer_token] = creditcard
+            post[:customer_token] = payment_method
           end
+        else
+          raise ArgumentError, "Invalid payment method type: #{payment_method.class}"
+        end
+      end
+
+      # Get a single use token from Pin for NT based payment methods
+      # This is used to create a charge using a network tokenized card
+      def get_single_use_token(payment_method, options)
+        return unless payment_method.is_a?(NetworkTokenizationCreditCard)
+
+        # binding.pry
+        post = {
+          type: 'network_token',
+          source: {
+            number: payment_method.number,
+            expiry_month: payment_method.month,
+            expiry_year: payment_method.year,
+            network_type: payment_method.source.to_s.gsub('_', ''),
+            cryptogram: payment_method.payment_cryptogram,
+            eci: payment_method.eci
+          }
+        }
+        begin
+          result = commit(:post, 'payment_sources', post, options)
```

Comment:
❓ Question:

I'm entirely sure, but perhaps this call could e wrapped into a multi-response object?